src/app/games/martian-chess/MartianChessState.ts

```diff
@@ -0,0 +1,132 @@
+import { Utils } from 'src/app/utils/utils';
+import { MartianChessMove } from './MartianChessMove';
+
+export class MartianChessPiece {
+
+    public static EMPTY: MartianChessPiece = new MartianChessPiece(0);
+    public static PAWN: MartianChessPiece = new MartianChessPiece(1);
+    public static DRONE: MartianChessPiece = new MartianChessPiece(2);
+    public static QUEEN: MartianChessPiece = new MartianChessPiece(3);
+
+    public static tryMerge(left: MartianChessPiece, right: MartianChessPiece): MGPOptional<MartianChessPiece> {
+        const firstValue: number = left.value;
+        const secondValue: number = right.value;
+        assert(firstValue !== 0 && secondValue !== 0, 'tryMerge cannot be called with empty pieces');
+        const totalValue: number = firstValue + secondValue;
+        if (totalValue === 2 || totalValue === 3) {
+            return MGPOptional.of(MartianChessPiece.from(totalValue));
+        } else {
+            return MGPOptional.empty();
+        }
+    }
+    private static from(value: number): MartianChessPiece {
+        switch (value) {
+            case MartianChessPiece.DRONE.value: return MartianChessPiece.DRONE;
+            default:
+                Utils.expectToBe(value, MartianChessPiece.QUEEN.value);
+                return MartianChessPiece.QUEEN;
+        }
+    }
+
+    private constructor(private readonly value: number) {}
+
+    public getValue(): number {
+        return this.value;
+    }
+}
+export class MartianChessState extends GameStateWithTable<MartianChessPiece> {
+
+    public static getInitialState(): MartianChessState {
+        const _: MartianChessPiece = MartianChessPiece.EMPTY;
+        const A: MartianChessPiece = MartianChessPiece.PAWN;
+        const B: MartianChessPiece = MartianChessPiece.DRONE;
+        const C: MartianChessPiece = MartianChessPiece.QUEEN;
+        const board: Table<MartianChessPiece> = [
+            [C, C, B, _],
+            [C, B, A, _],
+            [B, A, A, _],
+            [_, _, _, _],
+            [_, _, _, _],
+            [_, A, A, B],
+            [_, A, B, C],
+            [_, B, C, C],
+        ];
+        return new MartianChessState(board, 0, MGPOptional.empty());
+    }
+    public readonly captured: MGPMap<Player, number>;
+
+    public constructor(board: Table<MartianChessPiece>,
+                       turn: number,
+                       public readonly lastMove: MGPOptional<MartianChessMove> = MGPOptional.empty(),
+                       public readonly countDown: MGPOptional<number> = MGPOptional.empty(),
+                       captured?: MGPMap<Player, number>)
+    {
+        super(board, turn);
+        if (captured == null) {
+            captured = new MGPMap();
+            captured.set(Player.ZERO, 0);
+            captured.set(Player.ONE, 0);
+        }
+        captured.makeImmutable();
+        this.captured = captured;
+    }
+    public getPlayerTerritory(player: Player): [number, number] {
```

Comment:
si tu fais `getPlayerTerritory: Set<number>` (soit Set(0,1,2,3), soit Set(4,5,6,7)), tu peux faire `for (const y of getPlayerTerritory(player)` et `getPlayerTerritory.contains(y)`, c'qui est plus lisible non ? (que les yRange[0] et yRange[1])